core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java

```diff
@@ -367,182 +326,363 @@ private void loadRebalance() {
-        DEFAULT_COLUMN_TYPES[NewLineProtoParser.ENTITY_TYPE_STRING] = ColumnType.STRING;
-        DEFAULT_COLUMN_TYPES[NewLineProtoParser.ENTITY_TYPE_BOOLEAN] = ColumnType.BOOLEAN;
-        DEFAULT_COLUMN_TYPES[NewLineProtoParser.ENTITY_TYPE_LONG256] = ColumnType.LONG256;
+    void setListener(LineTcpReceiver.SchedulerListener listener) {
+        this.listener = listener;
+    }
+
+    private TableUpdateDetails startNewMeasurementEvent(NetworkIOJob netIoJob, LineTcpParser protoParser) {
+        final TableUpdateDetails tableUpdateDetails = netIoJob.getTableUpdateDetails(protoParser.getMeasurementName());
+        if (null != tableUpdateDetails) {
+            return tableUpdateDetails;
+        }
+        return startNewMeasurementEvent0(netIoJob, protoParser);
+    }
+
+    private TableUpdateDetails startNewMeasurementEvent0(NetworkIOJob netIoJob, LineTcpParser protoParser) {
+        TableUpdateDetails tableUpdateDetails;
+        tableUpdateDetailsLock.writeLock().lock();
+        try {
+            int keyIndex = tableUpdateDetailsByTableName.keyIndex(protoParser.getMeasurementName());
+            if (keyIndex < 0) {
+                tableUpdateDetails = tableUpdateDetailsByTableName.valueAt(keyIndex);
+            } else {
+                String tableName = protoParser.getMeasurementName().toString();
+                int status = engine.getStatus(securityContext, path, tableName, 0, tableName.length());
+                if (status != TableUtils.TABLE_EXISTS) {
+                    LOG.info().$("creating table [tableName=").$(tableName).$(']').$();
+                    engine.createTable(securityContext, ddlMem, path, tableStructureAdapter.of(tableName, protoParser));
+                }
+
+                keyIndex = idleTableUpdateDetailsByTableName.keyIndex(tableName);
+                if (keyIndex < 0) {
+                    LOG.info().$("idle table going active [tableName=").$(tableName).I$();
+                    tableUpdateDetails = idleTableUpdateDetailsByTableName.valueAt(keyIndex);
+                    idleTableUpdateDetailsByTableName.removeAt(keyIndex);
+                    tableUpdateDetailsByTableName.put(tableUpdateDetails.tableName, tableUpdateDetails);
+                } else {
+                    TelemetryTask.doStoreTelemetry(engine, Telemetry.SYSTEM_ILP_RESERVE_WRITER, Telemetry.ORIGIN_ILP_TCP);
+                    tableUpdateDetails = assignTableToWriterThread(tableName);
+                }
+            }
+
+            netIoJob.addTableUpdateDetails(tableUpdateDetails);
+            return tableUpdateDetails;
+        } finally {
+            tableUpdateDetailsLock.writeLock().unlock();
+        }
+    }
+
+    boolean tryButCouldNotCommit(NetworkIOJob netIoJob, LineTcpParser protoParser, FloatingDirectCharSink charSink) {
+        TableUpdateDetails tableUpdateDetails;
+        try {
+            tableUpdateDetails = startNewMeasurementEvent(netIoJob, protoParser);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *tableUpdateDetails*. Currently, there are execution paths that do not contain closure statements, for example, when an else condition at line 411 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).